docs/01-dynamic-type-tests.md

```diff
@@ -10,34 +10,255 @@ next: primitives.html
 In addition to `null` checks, there are several other dynamic type tests
 (*predicates*) on local variables that Flow recognizes and uses to refine
 types. Refining a type with a predicate means narrowing the original type with
-the type satisfied by values satisfying the predicate. Some of these
-predicates are:
+the type satisfied by values satisfying the predicate.
 
-- `typeof x == 'string'`, `typeof x == 'number'`, and so on: narrows the type
-  of `x` with `string`, `number`, and so on;
-- `x instanceof A`: narrows the type of `x` with `A`;
-- combinations of simpler predicates with logical operators `&&`, `||`, and
-  `!:` narrows the types of local variables appearing in the predicates with
-  types describing the union, intersection, or complement of values satisfying
-  those predicates.
+Type tests can occur in `if` and `switch` statements, the test block in loop
+constructs like `for`, `for-in`, `for-of`, and `do-while`, conditional
+expressions (ternary statements), and inline logical expressions like `a &&
+a.b`.
 
-For example, the following code typechecks:
+Flow understands many idiomatic constructs used to determine the type of a value
+at runtime, and incorporates that knowledge into its static analysis.
+
+### Maybe, Null, and Undefined
+
+{% highlight javascript linenos=table %}
+/* @flow */
+
+function foo(x: ?string): string {
+  if (x == null) {
+    return x;
+  } else {
+    return "null or undefined";
+  }
+}
+
+function bar(x: ?string): string {
+  if (x === null) {
+    return "null";
+  } else if (x === undefined) {
+    return "undefined";
+  } else {
+    return x;
+  }
+}
+{% endhighlight %}
+
+Read more about <a href="docs/nullable-types.html">Maybe Types</a>.
+
+### Truthiness and Existance
+
+{% highlight javascript linenos=table %}
+/* @flow */
+
+function foo(x: string): string {
+  if (x) {
+    return x;
+  } else {
+    return "the empty string";
+  }
+}
+
+function bar(x: number): string {
+  if (x) {
+    return x.toString();
+  } else {
+    return "zero";
+  }
+}
+
+function baz(x: boolean): string {
+  if (x) {
+    return "true";
+  } else {
+    return "false";
+  }
+}
```

Comment:
knowing a boolean is truthy is useful for nested logical operators like `x && "truthy" || "falsy"`. without knowing that, it'd think it could return `boolean | string` (false || "falsy", which is clearly just "falsy" to a human). but I agree, it's hard to find a user-level example without getting into internals that should appear to Just Work.